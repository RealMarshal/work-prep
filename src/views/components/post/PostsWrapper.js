import React from 'react'
import { connect } from 'react-redux'
import { lifecycle } from 'recompose'
import Typography from '@material-ui/core/Typography'
import CircularProgress from '@material-ui/core/CircularProgress'

import Post from './Post'
import { getPosts, totalPosts, totalAuthors } from '../../../state/ducks/posts/selectors'
import { fetchPosts } from '../../../state/ducks/posts/operations'


const PostsList = ({posts}) => (
  <div className='posts__postsList'>
    { !posts.data && !posts.error && <CircularProgress /> }
    { 
      posts.data && posts.data.map(post =>
        <Post key={post.id} title={post.title} text={post.body} /> )
    }
    { posts.error && <Typography className='post__errorMessage' variant='caption' color='error'>Ошибка получения данных</Typography> }
  </div>
)

const PostsStats = ({totalPosts, totalAuthors}) => (
  <div className='posts__postsStats'>
    <Typography className='post__totalPosts'>Постов { totalPosts }</Typography>
    <Typography className='post__totalAuthors'>Авторов { totalAuthors }</Typography>
  </div>
)

const PostsWrapper = ({posts, totalPosts, totalAuthors}) => (
  <div className='posts'>
    <PostsStats totalPosts={totalPosts} totalAuthors={totalAuthors} />
    <PostsList posts={posts} />
  </div>
)

const mapStateToProps = state => ({
  posts: getPosts(state),
  totalPosts: totalPosts(state),
  totalAuthors: totalAuthors(state)
})

const mapDispatchToProps = {
  fetchPosts
}

const withFetchPosts = lifecycle({
  componentDidMount() {
    this.props.fetchPosts()
  }
})

export { PostsWrapper, PostsStats, PostsList }
export default connect(mapStateToProps, mapDispatchToProps)(withFetchPosts(PostsWrapper))