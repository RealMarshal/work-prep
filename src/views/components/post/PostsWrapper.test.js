import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { PostsWrapper, PostsStats, PostsList } from './PostsWrapper'
import CircularProgress from '@material-ui/core/CircularProgress'

describe('PostsStats', () => {

  const testProps = {
    totalPosts: 1,
    totalAuthors: 1
  }
  const rendered = shallow(<PostsStats {...testProps} />)

  it('should render correctly', () => {
    expect(toJson(rendered)).toMatchSnapshot()
  })

  it('should show number of posts', () => {
    const expectedValue = `Постов ${testProps.totalPosts}`
    expect(rendered.find('.post__totalPosts').render().text()).toEqual(expectedValue)
  })

  it('should show number of authors', () => {
    const expectedValue = `Авторов ${testProps.totalAuthors}`
    expect(rendered.find('.post__totalAuthors').render().text()).toEqual(expectedValue)
  })
})

describe('PostsList', () => {

  it('should show loading indicator', () => {
    const testProps = {
      posts: []
    }
    const rendered = shallow(<PostsList {...testProps} />)
    expect(rendered.find(CircularProgress).exists()).toEqual(true)
  })

  it('should show posts', () => {
    const testProps = {
      posts: {
        data: [{ title: 'sdsd', body: 'sdsdwqe', id: 4}]
      }
    }
    const expectedPost = {
      title: testProps.posts.data[0].title,
      text: testProps.posts.data[0].body
    }
    const rendered = shallow(<PostsList {...testProps} />)
    expect(rendered.find('Post').props()).toEqual(expectedPost)
  })

  it('should show error', () => {
    const testProps = {
      posts: {
        error: 'Error happened'
      }
    }
    const rendered = shallow(<PostsList {...testProps} />)
    expect(rendered.find('.post__errorMessage').render().text()).toEqual('Ошибка получения данных')
  })
})

describe('PostsWrapper', () => {

  const testProps = {
    posts: [{ title: 'asdas', text: 'asdqwe', id: 3}],
    totalPosts: 1,
    totalAuthors: 1
  }
  const rendered = shallow(<PostsWrapper {...testProps} />)

  it('should render correctly', () => {
    expect(toJson(rendered)).toMatchSnapshot()
  })

  it('should contain PostsStats', () => {
    expect(rendered.find(PostsStats).exists()).toEqual(true)
  })

  it('should contain PostsList', () => {
    expect(rendered.find(PostsList).exists()).toEqual(true)
  })

  it('should pass stats to PostsStats', () => {
    const { totalPosts, totalAuthors } = testProps
    expect(rendered.find(PostsStats).props()).toEqual({totalPosts, totalAuthors})
  })

  it('should pass posts to PostsList', () => {
    const { posts } = testProps
    expect(rendered.find(PostsList).props()).toEqual({posts})
  })
})