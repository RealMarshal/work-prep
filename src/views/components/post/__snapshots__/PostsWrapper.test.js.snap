// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsStats should render correctly 1`] = `
<div
  className="posts__postsStats"
>
  <WithStyles(Typography)
    className="post__totalPosts"
  >
    Постов 
    1
  </WithStyles(Typography)>
  <WithStyles(Typography)
    className="post__totalAuthors"
  >
    Авторов 
    1
  </WithStyles(Typography)>
</div>
`;

exports[`PostsWrapper should render correctly 1`] = `
<div
  className="posts"
>
  <PostsStats
    totalAuthors={1}
    totalPosts={1}
  />
  <PostsList
    posts={
      Array [
        Object {
          "id": 3,
          "text": "asdqwe",
          "title": "asdas",
        },
      ]
    }
  />
</div>
`;
